## Answers to StageB Quiz

### Question 1

#### In Linear regression. L2 regularizaion is equivalent to imposing a:

#### Answer: Gaussian Prior

### Question 2

#### Cross_Validation:

#### Answer: is often use to select hyperparameters. is gauranteed to prevent overfitting

### Question 3

#### Ridge Regression

#### Answer: Reduces variance at the expenses of higher bias

### Question 4

#### In the diffrent terms of the bias-variance tradeoff, which of the following is substantially more hamrful to the test error than the training error?

#### Answer: Variance

### Question 5

#### What can you use to find the best fit line for data in Linear Regression

#### Answer: least square error

### Question 6

#### Which of the following sentence is true about outliers in Linear Regression

#### Answer: Linear Regression is sensitive to outliers

### Question 7

#### How many coefficients do you need to estimate a simple linear regression model (one independet variable)?

#### Answer: 2

### Question 8

#### Adding more basis functions in a linear model:

#### Answer: Decreases model bias

### Question 9

#### A best fit line relating X and Y has a R-Squared value of 0.75. How do i interpret this information?

#### Answer: The model is performing with 75% accuracy

### Question 10

#### The Lasso can be interpreted as least-squares linear regression where:

#### Answer: Weights are regularized with the L1 norm

### Question 11

#### Which of these is not an assumption of Linear Regression

#### Answer: Heteroscedacity

### Question 12

#### From the dataset, fit a linear model on the relationship between the temperature in the living room in celsiun (x=T2), and the temperature outside the building(y=T6), what is the R^2 value in 2 D.P?

#### Answer: 0.64

## Codes shown below

## Importing all important libriaries

In [5]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.preprocessing import MinMaxScaler



In [2]:
df = pd.read_csv('energydata_complete.csv')

In [3]:
df.head().T

0                    1                    2  \
date         2016-01-11 17:00:00  2016-01-11 17:10:00  2016-01-11 17:20:00   
Appliances                    60                   60                   50   
lights                        30                   30                   30   
T1                         19.89                19.89                19.89   
RH_1                     47.5967              46.6933                 46.3   
T2                          19.2                 19.2                 19.2   
RH_2                       44.79              44.7225              44.6267   
T3                         19.79                19.79                19.79   
RH_3                       44.73                44.79              44.9333   
T4                            19                   19              18.9267   
RH_4                     45.5667              45.9925                45.89   
T5                       17.1667              17.1667              17.1667   
RH_5                        55.2                 55.2                55.09   
T6                       7.02667              6.83333                 6.56   
RH_6                     84.2567              84.0633              83.1567   
T7                          17.2                 17.2                 17.2   
RH_7                     41.6267                41.56              41.4333   
T8                          18.2                 18.2                 18.2   
RH_8                        48.9              48.8633                48.73   
T9                       17.0333              17.0667                   17   
RH_9                       45.53                45.56                 45.5   
T_out                        6.6              6.48333              6.36667   
Press_mm_hg                733.5                733.6                733.7   
RH_out                        92                   92                   92   
Windspeed                      7              6.66667              6.33333   
Visibility                    63              59.1667              55.3333   
Tdewpoint                    5.3                  5.2                  5.1   
rv1                      13.2754              18.6062              28.6427   
rv2                      13.2754              18.6062              28.6427   

                               3                    4  
date         2016-01-11 17:30:00  2016-01-11 17:40:00  
Appliances                    50                   60  
lights                        40                   40  
T1                         19.89                19.89  
RH_1                     46.0667              46.3333  
T2                          19.2                 19.2  
RH_2                       44.59                44.53  
T3                         19.79                19.79  
RH_3                          45                   45  
T4                         18.89                18.89  
RH_4                     45.7233                45.53  
T5                       17.1667                 17.2  
RH_5                       55.09                55.09  
T6                       6.43333              6.36667  
RH_6                     83.4233              84.8933  
T7                       17.1333                 17.2  
RH_7                       41.29                41.23  
T8                          18.1                 18.1  
RH_8                       48.59                48.59  
T9                            17                   17  
RH_9                        45.4                 45.4  
T_out                       6.25              6.13333  
Press_mm_hg                733.8                733.9  
RH_out                        92                   92  
Windspeed                      6              5.66667  
Visibility                  51.5              47.6667  
Tdewpoint                      5                  4.9  
rv1                      45.4104              10.0841  
rv2                      45.4104              10.0841

In [6]:
# Drop the date and lights columns
df.drop(['date', 'lights'], axis=1, inplace=True)

In [8]:
# Then we normalize our dataset to a normal scale using the MinMaxscaler
scaler = MinMaxScaler()
normalised_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

In [12]:
# Check the head of our normalized dataset
normalised_df.head().T

0         1         2         3         4
Appliances   0.046729  0.046729  0.037383  0.037383  0.046729
T1           0.327350  0.327350  0.327350  0.327350  0.327350
RH_1         0.566187  0.541326  0.530502  0.524080  0.531419
T2           0.225345  0.225345  0.225345  0.225345  0.225345
RH_2         0.684038  0.682140  0.679445  0.678414  0.676727
T3           0.215188  0.215188  0.215188  0.215188  0.215188
RH_3         0.746066  0.748871  0.755569  0.758685  0.758685
T4           0.351351  0.351351  0.344745  0.341441  0.341441
RH_4         0.764262  0.782437  0.778062  0.770949  0.762697
T5           0.175506  0.175506  0.175506  0.175506  0.178691
RH_5         0.381691  0.381691  0.380037  0.380037  0.380037
T6           0.381070  0.375443  0.367487  0.363800  0.361859
RH_6         0.841827  0.839872  0.830704  0.833401  0.848264
T7           0.170594  0.170594  0.170594  0.164310  0.170594
RH_7         0.653428  0.651064  0.646572  0.641489  0.639362
T8           0.173329  0.173329  0.173329  0.164175  0.164175
RH_8         0.661412  0.660155  0.655586  0.650788  0.650788
T9           0.223032  0.226500  0.219563  0.219563  0.219563
RH_9         0.677290  0.678532  0.676049  0.671909  0.671909
T_out        0.372990  0.369239  0.365488  0.361736  0.357985
Press_mm_hg  0.097674  0.100000  0.102326  0.104651  0.106977
RH_out       0.894737  0.894737  0.894737  0.894737  0.894737
Windspeed    0.500000  0.476190  0.452381  0.428571  0.404762
Visibility   0.953846  0.894872  0.835897  0.776923  0.717949
Tdewpoint    0.538462  0.533937  0.529412  0.524887  0.520362
rv1          0.265449  0.372083  0.572848  0.908261  0.201611
rv2          0.265449  0.372083  0.572848  0.908261  0.201611

## Code shown below

In [13]:
# firstly, we select T2 and T6 columns as X and y
X = normalised_df['T2']
y = normalised_df['T6']

In [14]:
# We reshape our X column to a two dimensional shape
X = np.array([X]).reshape(-1, 1)

In [15]:
# Now, we split our dataset into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [16]:
model = LinearRegression()

In [17]:
# We fit the model to the training dataset
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [23]:
# Next, we make some predictions
y_pred = model.predict(X_test)

In [24]:
# R^2 value
print(f'The R^2 value is:{(r2_score(y_test, y_pred).round(2))}')

The R^2 value is:0.64


### Question 13

#### What is the Mean Absolute Error(in two decimal places)?

#### Answer: 0.05

## Codes shown below

In [26]:
# Here we drop the Appliances column, which is use as the target variable
features_df = normalised_df.drop('Appliances', axis=1)
target_df = normalised_df['Appliances']

In [28]:
# Now, we split our dataset into training and test set
X_train, X_test, y_train, y_test = train_test_split(features_df, target_df, test_size=0.3, random_state=42)

In [29]:
model_1 = LinearRegression()

In [30]:
model_1.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [31]:
y_predictions = model_1.predict(X_test)

In [33]:
# Mean Absolute Error
print(f'The MAE value is: {mean_absolute_error(y_test, y_predictions).round(2)}')

The MAE value is: 0.05


### Question 14

#### What is the Residual Sum of Squares(in two decimal places)?

#### Answer: 45.35


## Codes shown below

In [34]:
# RSS
print(f'The RSS value is: {np.sum(np.square(y_test - y_predictions)).round(2)}')

The RSS value is: 45.35


### Question 15

#### What is the Root Mean Square Error in (Three decimal places)?

#### Answer: 0.088

## codes shown below

In [36]:
print(f'The RMSE value is: {np.sqrt(mean_squared_error(y_test, y_predictions)).round(3)}')

The RMSE value is: 0.088


### Question 16

#### What is the Coefficient of Determination (in Two decimal places)?

#### Answer: 0.15

## codes shown below

In [38]:
print(f'The Coefficient of Determination is: {r2_score(y_test, y_predictions).round(2)}')

The Coefficient of Determination is: 0.15


### Question 17

#### Obtain the feature weights from your linear model above. Which feature has the lowest and highest weights respectively?

#### Answer: RH_2, RH_1

## codes shown below

In [61]:
def feature_weights_df(model_1, feat, col_name):
    '''
    This function gets all the models features weight defined and 
    computes their weights.
    
    Parameters: model_1, feat, col_name.
    '''
    # Define a Series of sort values of all features from lowest to highest
    weights = pd.Series(model_1.coef_, feat.columns).sort_values()
    # Create a DataFrame of the features
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

In [65]:
linear_model = feature_weights_df(model_1, X_train, 'Linear_Model')
linear_model

Features  Linear_Model
0          RH_2     -0.456698
1         T_out     -0.321860
2            T2     -0.236178
3            T9     -0.189941
4          RH_8     -0.157595
5        RH_out     -0.077671
6          RH_7     -0.044614
7          RH_9     -0.039800
8            T5     -0.015657
9            T1     -0.003281
10          rv1      0.000770
11          rv2      0.000770
12  Press_mm_hg      0.006839
13           T7      0.010319
14   Visibility      0.012307
15         RH_5      0.016006
16         RH_4      0.026386
17           T4      0.028981
18    Windspeed      0.029183
19         RH_6      0.038049
20         RH_3      0.096048
21           T8      0.101995
22    Tdewpoint      0.117758
23           T6      0.236425
24           T3      0.290627
25         RH_1      0.553547

#### RH_2 has a weight value of -0.456698 to make it the lowest, while RH_1 has a weight value of 0.553547 to make it the highest

### Question 18

#### Train a ridge regression model with an aplha value of 0.4. is there any change to the root mean square error (RMSE) when evaluated on the test set?

#### Answer: No

## codes shown below

In [86]:
# Here we train and predict the Ridge model with an alpha value set to 0.4
ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(X_train, y_train)
ridge_pred = ridge_reg.predict(X_test)

In [87]:
r2_score(y_test, ridge_pred)

0.1485249354509257

### Question 19

#### Train a Lasso regression model with an aplha value of 0.001. and obtain the new feature weights with it. How many of the features have non-zero feature weights?

#### Answer: 4

## codes shown below

In [88]:
# Here we train a lasso model with an alpha value set to 0.001
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)
lasso_pred = lasso_reg.predict(X_test)


In [89]:
lasso_reg = feature_weights_df(lasso_reg, X_train, 'Lasso_Model')
lasso_reg

Features  Lasso_Model
0        RH_out    -0.049557
1          RH_8    -0.000110
2            T1     0.000000
3     Tdewpoint     0.000000
4    Visibility     0.000000
5   Press_mm_hg    -0.000000
6         T_out     0.000000
7          RH_9    -0.000000
8            T9    -0.000000
9            T8     0.000000
10         RH_7    -0.000000
11          rv1    -0.000000
12           T7    -0.000000
13           T6     0.000000
14         RH_5     0.000000
15           T5    -0.000000
16         RH_4     0.000000
17           T4    -0.000000
18         RH_3     0.000000
19           T3     0.000000
20         RH_2    -0.000000
21           T2     0.000000
22         RH_6    -0.000000
23          rv2    -0.000000
24    Windspeed     0.002912
25         RH_1     0.017880

### Question 20

#### What is the new RMSE with the lasso regression (in 3 decimal places)?

#### Answer: 0.094

## codes shown below

In [90]:
print(f'The new RMSE value with lasso regression is: {np.sqrt(mean_squared_error(y_test, lasso_pred )).round(3)}')

The new RMSE value with lasso regression is: 0.094


## THANK YOU!!!